drivers/clickhouse/clickhouse.go

```diff
@@ -35,7 +38,82 @@ func init() {
 			}
 			return false
 		},
-		Copy:              drivers.CopyWithInsert(func(int) string { return "?" }),
+		Copy:              CopyWithInsert,
 		NewMetadataReader: NewMetadataReader,
 	})
 }
+
+// CopyWithInsert builds a copy handler based on insert.
+func CopyWithInsert(ctx context.Context, db *sql.DB, rows *sql.Rows, table string) (int64, error) {
+	columns, err := rows.Columns()
+	if err != nil {
+		return 0, fmt.Errorf("failed to fetch source rows columns: %w", err)
+	}
+	clen := len(columns)
+	query := table
+	if !strings.HasPrefix(strings.ToLower(query), "insert into") {
+		leftParen := strings.IndexRune(table, '(')
+		if leftParen == -1 {
+			colRows, err := db.QueryContext(ctx, "SELECT * FROM "+table+" WHERE 1=0")
+			if err != nil {
+				return 0, fmt.Errorf("failed to execute query to determine target table columns: %w", err)
+			}
+			columns, err := colRows.Columns()
+			_ = colRows.Close()
+			if err != nil {
+				return 0, fmt.Errorf("failed to fetch target table columns: %w", err)
+			}
+			table += "(" + strings.Join(columns, ", ") + ")"
+		}
+		// TODO if the db supports multiple rows per insert, create batches of 10 rows
+		query = "INSERT INTO " + table + " VALUES (" + strings.Repeat("?, ", clen-1) + "?)"
+	}
+	tx, err := db.BeginTx(ctx, nil)
+	if err != nil {
+		return 0, fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	stmt, err := tx.PrepareContext(ctx, query)
+	if err != nil {
+		return 0, fmt.Errorf("failed to prepare insert query: %w", err)
+	}
+	defer stmt.Close()
+	columnTypes, err := rows.ColumnTypes()
+	if err != nil {
+		return 0, fmt.Errorf("failed to fetch source column types: %w", err)
+	}
+	values := make([]interface{}, clen)
+	valueRefs := make([]reflect.Value, clen)
+	actuals := make([]interface{}, clen)
+	for i := 0; i < len(columnTypes); i++ {
+		valueRefs[i] = reflect.New(columnTypes[i].ScanType())
+		values[i] = valueRefs[i].Interface()
+	}
+	var n int64
+	for rows.Next() {
+		err = rows.Scan(values...)
+		if err != nil {
+			return n, fmt.Errorf("failed to scan row: %w", err)
+		}
+		//We can't use values... in Exec() below, because, in some cases, clickhouse
```

Comment:
Is this something that others drivers would also benefit from?

Reply:
Yes. I think both changes apply to the general drivers.CopyWithInsert implementation. More testing will be needed though since CopyWithInsert is used in 10 drivers.

Reply:
How about doing this in three PRs

1. This PR fixes Clickhouse only at first
2. I do https://github.com/xo/usql/issues/458 for Clickhouse
3. Replace drivers.CopyWithInsert with this implementation + optional batching + the placeholder code currently in drivers.CopyWithInsert . Extend TestCopy drivers_test.go to include a source DB that doesn't support ScanType (e.g. csvq), and a target DB that only support prepared INSERT statements (e.g. Clickhouse, Databend).

Trying to go directly to step 3 will delay the fix for \copy for Clickhouse a lot .